pkg/lib/operation/template/local/test/operation.go

```diff
@@ -147,10 +172,27 @@ func runSingleTest(testName string, tmpl *template.Template, _ bool, repoDirFS f
 	}
 	d.Logger().Debugf(`Inputs prepared.`)
 
+	// Pull the project first in E2E test
+	if !localTest {
+		pullOpts := syncPull.Options{
+			DryRun:            false,
+			LogUntrackedPaths: true,
+		}
+		err = syncPull.Run(prjState, pullOpts, projectDeps)
+		if err != nil {
+			return err
+		}
+
+		prjState, err = prj.LoadState(loadState.LocalOperationOptions())
+		if err != nil {
+			return err
+		}
+	}
+
```

Comment:
@webrouse Potřeboval bych tady trochu pomoct. Když chci (níže) zavolat push, tak musím nejdřív udělat pull. Nebo možná nemusím, ale je potřeba nějak syncnout že v tom remote projektu je main branch. Když jsem měl jen use a push, tak mi to failovalo na tom, že chtěl v remote vytvářet branch a to samozřejmě nešlo.

Tak jsem přidal pull, ale začalo to failovat tady že `branchState.Local` je `nil`:
https://github.com/keboola/keboola-as-code/blob/bd717876bb5d90b87287610b4cfe5dbdc028a1ea/pkg/lib/operation/project/local/template/use/operation.go#L121-L121

Tak jsem přidal tady ještě jednou `prj.LoadState()` aby se po tom pullu aktualizoval. Jenže teď mi to zase failuje tady na `Unexpected panic: interface conversion: model.ObjectState is nil, not *model.BranchState`:
https://github.com/keboola/keboola-as-code/blob/bd717876bb5d90b87287610b4cfe5dbdc028a1ea/pkg/lib/operation/project/local/template/use/operation.go#L113-L113

Tak se chci ujistit jestli vůbec postupuju dobře a nebyla by na to lepší cesta.

Reply:
idem pozriet

Reply:
@JakubMatejka pride mi, ze ten remote test je velmi podobny `use template`, ktory mame v API nie?
https://github.com/keboola/keboola-as-code/blob/bd90aee12f20dbae825b235fa526b85e001c1507/internal/pkg/api/server/templates/service/service.go#L128-L136

Tam je to samozrejme nahackovane, ale funkcne, neda sa z toho vychadzat?
Co su rozdiely/problemy, ktore potrebujes vyriesit?

Reply:
Mozno by bolo pre teba jednoduchsie riesit local/remote test oddelene a az nakoniec vytiahnut spolocnu cast kodu mimo.
Ono to bude cele jednoduchsie po refacotringu, len musime chvilu este vydrzat :face_with_spiral_eyes:

Reply:
Ok, rozdělil jsem to na `runLocalTest` a `RunRemoteTest` a zkusil jsem do `RunRemoteTest` přidat co bylo v tom `service. UseTemplateVersion`, manifest (https://github.com/keboola/keboola-as-code/blob/63d3d42317abf465ee73b89d9231d178c6f64ed8/pkg/lib/operation/template/local/test/operation.go#L235-L237) a zkopírování stavu namísto zavolání pull (https://github.com/keboola/keboola-as-code/blob/63d3d42317abf465ee73b89d9231d178c6f64ed8/pkg/lib/operation/template/local/test/operation.go#L277-L283)

Jenže teď mi to zase failuje na tom, že v názvu konfigu jsou nějaký čísla a já nevím kde se tam vzaly. 😅 Vypadá to jako idčko ale v namingu není napsaný že by tam mělo být. 🤔 
![CleanShot 2022-08-18 at 20 54 26@2x](https://user-images.githubusercontent.com/215660/185472725-c89dc962-792f-4284-8fd7-3221fdb9a5ae.jpg)

Reply:
Tak to si na tom uz celkom dobre, nie ? :) 

> ale v namingu není napsaný že by tam mělo být
Mas tam 2 metody v `naming`:
- `TemplateWithIds` / `TemplateWithoutIds`

Verzis s IDs je pouzita ako default hodnota tu:
https://github.com/keboola/keboola-as-code/blob/bd90aee12f20dbae825b235fa526b85e001c1507/internal/pkg/project/manifest/file.go#L47

A ked tam pouzivas niekde fake manifest, tak ta hodnota sa asi berie odtial ^^^

Reply:
> Tak to si na tom uz celkom dobre, nie ? :)

To už si říkám pár dní. 😄

> Mas tam 2 metody v `naming`:

Aha kua, já si to popletl s tím že to mám v namingu v tom manifestu (https://github.com/keboola/keboola-as-code/blob/63d3d42317abf465ee73b89d9231d178c6f64ed8/test/cli/template-test/remote/out/my-template/v0/tests/two/expected-out/.keboola/manifest.json#L10-L10) ale to je expected state. 🙈 Díky, zkusím.

Reply:
Aha ne, zpět. 🙂 Ten manifest ze kterýcho vznikne actual state je ve fixtures: https://github.com/keboola/keboola-as-code/blob/63d3d42317abf465ee73b89d9231d178c6f64ed8/internal/pkg/fixtures/local/empty-branch/.keboola/manifest.json#L18-L18

Tady ho načtu do `projectFS`: https://github.com/keboola/keboola-as-code/blob/63d3d42317abf465ee73b89d9231d178c6f64ed8/pkg/lib/operation/template/local/test/operation.go#L215-L215
A tady z něj sestavím `Project`: https://github.com/keboola/keboola-as-code/blob/63d3d42317abf465ee73b89d9231d178c6f64ed8/pkg/lib/operation/template/local/test/operation.go#L233-L233 - to by mělo použít ten manifest z FS (a jeho naming), ne? 

Přijde mi že by to mělo dělat to samý jako když ta service vytvoří fake manifest a potom z něj vytvoří ten `Project`:
https://github.com/keboola/keboola-as-code/blob/63d3d42317abf465ee73b89d9231d178c6f64ed8/internal/pkg/project/project.go#L45-L60

Reply:
Idem to pozriet, myslel som, ze tam pouzivas fake manifest (nepozeral som do kodu), potom by davala zmysel ta default hodnota ...

Reply:
Stale na to pozeram, mas pravdu, ze tam na prvy pohlad nie je vidiet dovod, preco by to nemalo ist.

Reply:
@JakubMatejka ...a je ten testovaci projekt prazdny? Skor mi to pride ako poradove cislo, tj. `empty-007` znamena, ze je to 8-ma konfiguracia s nazvom `"empty"`.

Reply:
omg, asi není. 🙈 Jsem nějak latentně myslel že z `testproject.GetTestProject()` dostanu automaticky prázdný projekt, ale to je vlastně blbost. Tak díky. 😅

Reply:
jj, to riesi az `SetState`, podla toho aky stav projektu potrebujes.

Reply:
hmm, zkusil jsem https://github.com/keboola/keboola-as-code/pull/767/commits/ab0a06b253b4652a39c6a5ee6544f34425719c35 a nempomáhá to. 😕 Dívám se jak je to `SetState` použitý jinde a nic co by tam chybělo tam nevidím.

Reply:
Ale kua, jen na tom jednom konfigu `empty`. 🤔 Není tam sakra někde už na začátku nějaký jiný se stejným jménem. 
![CleanShot 2022-08-19 at 14 37 54@2x](https://user-images.githubusercontent.com/215660/185620188-8a640ba1-4a70-49b1-88b8-100dab679c40.jpg)

Reply:
Ale v local fixture `empty-branch` pro FS (https://github.com/keboola/keboola-as-code/blob/ab0a06b253b4652a39c6a5ee6544f34425719c35/internal/pkg/fixtures/local/empty-branch/.keboola/manifest.json#L1-L1) ani v remote fixture `empty` pro ten SetState (https://github.com/keboola/keboola-as-code/blob/ab0a06b253b4652a39c6a5ee6544f34425719c35/internal/pkg/fixtures/remote/empty.json#L1-L1) nic není. 😕

Reply:
@JakubMatejka no ale minimal ma prave ten `empty` config v sebe: https://github.com/keboola/keboola-as-code/blob/main/internal/pkg/fixtures/remote/minimal.json

Reply:
Chces asi `empty`: https://github.com/keboola/keboola-as-code/blob/main/internal/pkg/fixtures/remote/empty.json

Reply:
(kludne ten minimal premenuj v celom projekte na nieco ine ak to je zly nazov)

Reply:
🤦‍♂️ omg, já se dívám do empty, myslím si že používám empty a mám tam minimal. Ach jó, omlouvám se. Už jsem asi zase přehřátej. Vyrazil jsem po čase do kanclu a je tu na chcípnutí ještě víc než doma. 😅

Reply:
V pohode :)